oximeter/instruments/src/kstat/sampler.rs

```diff
@@ -0,0 +1,688 @@
+    // kstats at their intervals. Samples will be pushed onto the queue.
+    async fn run(mut self) {
+        // The collection of futures which yield IDs of targets to be sampled,
+        // after their interval expires. We need to push a really long future
+        // into this set, otherwise the `.next()` method will always return
+        // `None`.
+        let mut sample_timeouts = FuturesUnordered::new();
+        const DUMMY_FUT_DURATION: Duration = Duration::from_secs(60 * 60 * 24);
+        sample_timeouts
+            .push(YieldIdAfter::new(TargetId(0), DUMMY_FUT_DURATION));
+
+        loop {
+            tokio::select! {
+                maybe_id = sample_timeouts.next() => {
+                    let Some((id, interval)) = maybe_id else {
+                        warn!(
+                            self.log,
+                            "sample timeout future yielded None \
+                            requeueing long dummy future";
+                            "duration" => ?DUMMY_FUT_DURATION,
+                        );
+                        sample_timeouts.push(YieldIdAfter::new(
+                            TargetId(0),
+                            DUMMY_FUT_DURATION,
+                        ));
+                        continue;
+                    };
+                    match self.sample_one(id) {
+                        Ok(Some(mut samples)) => {
+                            debug!(
+                                self.log,
+                                "pulled samples from target";
+                                "id" => ?id,
+                                "n_samples" => samples.len(),
+                            );
+
+                            // Limit the number of samples we actually contain
+                            // in the sample queue. This is to avoid huge
+                            // allocations when we produce lots of data, but
+                            // we're not polled quickly enough by oximeter.
+                            let mut current_samples = self.samples.lock().unwrap();
+                            let n_new_samples = samples.len();
+                            let n_current_samples = current_samples.len();
+                            let n_total_samples = n_new_samples + n_current_samples;
+                            let n_overflow_samples =
+                                n_total_samples.checked_sub(self.sample_limit).unwrap_or(0);
+                            if n_overflow_samples > 0 {
+                                warn!(
+                                    self.log,
+                                    "sample queue is too full, dropping oldest samples";
+                                    "n_new_samples" => n_new_samples,
+                                    "n_current_samples" => n_current_samples,
+                                    "n_overflow_samples" => n_overflow_samples,
+                                );
```

Comment:
Great question, and no, I don't think it will be easily visible anywhere outside this log right now. One would need to infer it from a gap in the samples for a particular timeseries, or be looking at this log file. Those are both terrible, and generating our own timeseries to track these drops seems like a very good idea to surface it. I'll experiment with publishing that as well, since we'll mostly want to think about what fields to attach to such a timeseries.

Another point this raises for me is that noisy producers can negatively impact well-behaved ones, since they share the same sample queue. That interaction is not good. I think I have a reasonable idea for how to implement that in a way that doesn't complicate the locking story too much.

Thanks for bringing this up!

Reply:
Ok, I've done the per-target queue work in b6eb2dc95. I am going to go back and experiment with publishing our own self-stats here next.